vioscsi/vioscsi.c

```diff
@@ -462,36 +462,44 @@ ENTER_FN();
         return SP_RETURN_NOT_FOUND;
     }
 
-    /* Set num_queues and seg_max to some sane values, to keep "Static Driver Verification" happy */
+    /* Initialize the following variables to temporary values to keep
+     * Static Driver Verification happy. These values will be immediately
+     * reconfigured within GetScsiConfig below, which will retrieve the runtime
+     * values advertised by the underlying device.
+     */
     adaptExt->scsi_config.num_queues = 1;
-    adaptExt->scsi_config.seg_max = MAX_PHYS_SEGMENTS + 1;
+    adaptExt->scsi_config.seg_max    = SCSI_MINIMUM_PHYSICAL_BREAKS;
+    adaptExt->indirect               = FALSE;
     GetScsiConfig(DeviceExtension);
     SetGuestFeatures(DeviceExtension);
 
-    if(!adaptExt->dump_mode) {
-        adaptExt->indirect = CHECKBIT(adaptExt->features, VIRTIO_RING_F_INDIRECT_DESC);
-    }
-
-    ConfigInfo->NumberOfBuses               = 1;//(UCHAR)adaptExt->num_queues;
+    ConfigInfo->NumberOfBuses               = 1;
     ConfigInfo->MaximumNumberOfTargets      = min((UCHAR)adaptExt->scsi_config.max_target, 255/*SCSI_MAXIMUM_TARGETS_PER_BUS*/);
     ConfigInfo->MaximumNumberOfLogicalUnits = min((UCHAR)adaptExt->scsi_config.max_lun, SCSI_MAXIMUM_LUNS_PER_TARGET);
+    ConfigInfo->MaximumTransferLength       = SP_UNINITIALIZED_VALUE;       // Unlimited
+    ConfigInfo->NumberOfPhysicalBreaks      = SCSI_MINIMUM_PHYSICAL_BREAKS; // 16
 
-    if(adaptExt->dump_mode) {
-        ConfigInfo->NumberOfPhysicalBreaks  = SCSI_MINIMUM_PHYSICAL_BREAKS;
-    } else {
-        adaptExt->max_physical_breaks = PHYS_SEGMENTS;
-        if (adaptExt->indirect) {
-            adaptExt->max_physical_breaks = MAX_PHYS_SEGMENTS;
-            VioScsiReadRegistry(DeviceExtension);
+    if (!adaptExt->dump_mode) {
+        adaptExt->max_physical_breaks = adaptExt->indirect ? MAX_PHYS_SEGMENTS : PHYS_SEGMENTS;
+
+        /* Allow user to override max_physical_breaks via reg key
+         * [HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Services\vioscsi\Parameters\Device]
+         * "PhysicalBreaks"={dword value here}
+         */
+        VioScsiReadRegistryPhysicalBreaks(DeviceExtension);
+
+        if (adaptExt->scsi_config.max_sectors > 0 &&
+            adaptExt->scsi_config.max_sectors != 0xFFFF &&
+            adaptExt->max_physical_breaks * PAGE_SIZE > adaptExt->scsi_config.max_sectors * SECTOR_SIZE) {
+
+            adaptExt->max_physical_breaks = adaptExt->scsi_config.max_sectors * SECTOR_SIZE / PAGE_SIZE;
```

Comment:
Again, instead of all the clumsy conditions, it should really just be analogous to what's in `VioScsiReadRegistryPhysicalBreaks()`. `!= 0xFFFF` is just bad, and `> 0` is more or less wrong.

Is there any "verification" that would be "unhappy" if a local variable is introduced?

Reply:
Why is 0xFFFF bad? That's 65535 IIRC, which is hardcoded in qemu unless the user overrides it with their own value. 

I dont think >0 is wrong, though we could probably drop the > 0 as you'd probably have to try hard to advertise max_sectors as 0, I don't even know if you could do that :) though I have a feeling that _**someone**_ out there will end up filing a bug report saying that they indeed set max_sectors to zero for some reason

RE Variable - I dont think the static analyzer cares about such a variable

Anyhow, you are right, we could use the clamp snippet you provided before, it would get us to the same place.

Here's a logical walkthrough for the sake of doing it long hand
```
int mbp_ms_clamp = adaptExt->scsi_config.max_sectors * SECTOR_SIZE / PAGE_SIZE; 
    // mbp_ms_clamp = 65535 * 512 / 4096 = 8191.875
adaptExt->max_physical_breaks = min(max(SCSI_MINIMUM_PHYSICAL_BREAKS, mbp_ms_clamp), adaptExt->max_physical_breaks);
    // max_physical_breaks min(max(16, 8191.875), 512) = 512 -> good
```

Then there is our case where max_sectors = 2048 (1MB)
```
int mbp_ms_clamp = adaptExt->scsi_config.max_sectors * SECTOR_SIZE / PAGE_SIZE; 
    // mbp_ms_clamp = 2048 * 512 / 4096 = 256
adaptExt->max_physical_breaks = min(max(SCSI_MINIMUM_PHYSICAL_BREAKS, mbp_ms_clamp), adaptExt->max_physical_breaks);
    // max_physical_breaks min(max(16, 256), 512) = 256 -> good
```

Reply:
What I mean is, instead of 0, we should have the same lower limit for this as the registry clamp (i.e. SCSI_MINIMUM_PHYSICAL_BREAKS)

As for `!= 0xffff`, for one the maximum qemu can take could change; regardless of whether that would actually happen, so `!=` (instead of `>`) is bad. What I mean by "wrong" is, similar to the case of the lower limit, it's disgusting to have a clamp somewhere (registry), and to have a override somewhere else (here). I think the least we can do is to have some consistency. (For the record, there's nothing wrong with an override itself; that's purely a matter of preference.)